File: bigseqkit/grep.go
package bigseqkit

import (
	"ignis/driver/api"
	"ignis/executor/api/ipair"
	"strconv"
)

type SeqKitGrepOptions struct {
	inner GrepOptions
}

type GrepOptions struct {
	Config             KitConfig
	Pattern            *[]string
	PatternFile        *string
	UseRegexp          *bool
	DeleteMatched      *bool
	InvertMatch        *bool
	ByName             *bool
	BySeq              *bool
	OnlyPositiveStrand *bool
	MaxMismatch        *int
	IgnoreCase         *bool
	Degenerate         *bool
	Region             *string
	Circular           *bool
	Count              *bool
}

func (this *GrepOptions) setDefaults() *GrepOptions {
	this.Config.setDefaults()
	setDefault(&this.Pattern, []string{""})
	setDefault(&this.PatternFile, "")
	setDefault(&this.UseRegexp, false)
	setDefault(&this.DeleteMatched, false)
	setDefault(&this.InvertMatch, false)
	setDefault(&this.ByName, false)
	setDefault(&this.BySeq, false)
	setDefault(&this.OnlyPositiveStrand, false)
	setDefault(&this.MaxMismatch, 0)
	setDefault(&this.IgnoreCase, false)
	setDefault(&this.Degenerate, false)
	setDefault(&this.Region, "")
	setDefault(&this.Circular, false)
	setDefault(&this.Count, false)

	return this
}

func (this *SeqKitGrepOptions) Config(v *SeqKitConfig) *SeqKitGrepOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitGrepOptions) Pattern(v []string) *SeqKitGrepOptions {
	this.inner.Pattern = &v
	return this
}

func (this *SeqKitGrepOptions) PatternFile(v string) *SeqKitGrepOptions {
	this.inner.PatternFile = &v
	return this
}

func (this *SeqKitGrepOptions) UseRegexp(v bool) *SeqKitGrepOptions {
	this.inner.UseRegexp = &v
	return this
}

func (this *SeqKitGrepOptions) DeleteMatched(v bool) *SeqKitGrepOptions {
	this.inner.DeleteMatched = &v
	return this
}

func (this *SeqKitGrepOptions) InvertMatch(v bool) *SeqKitGrepOptions {
	this.inner.InvertMatch = &v
	return this
}

func (this *SeqKitGrepOptions) ByName(v bool) *SeqKitGrepOptions {
	this.inner.ByName = &v
	return this
}

func (this *SeqKitGrepOptions) BySeq(v bool) *SeqKitGrepOptions {
	this.inner.BySeq = &v
	return this
}

func (this *SeqKitGrepOptions) OnlyPositiveStrand(v bool) *SeqKitGrepOptions {
	this.inner.OnlyPositiveStrand = &v
	return this
}

func (this *SeqKitGrepOptions) MaxMismatch(v int) *SeqKitGrepOptions {
	this.inner.MaxMismatch = &v
	return this
}

func (this *SeqKitGrepOptions) IgnoreCase(v bool) *SeqKitGrepOptions {
	this.inner.IgnoreCase = &v
	return this
}

func (this *SeqKitGrepOptions) Degenerate(v bool) *SeqKitGrepOptions {
	this.inner.Degenerate = &v
	return this
}

func (this *SeqKitGrepOptions) Region(v string) *SeqKitGrepOptions {
	this.inner.Region = &v
	return this
}

func (this *SeqKitGrepOptions) Circular(v bool) *SeqKitGrepOptions {
	this.inner.Circular = &v
	return this
}

func commonGrep(input *api.IDataFrame[string], opts *GrepOptions) (*api.IDataFrame[string], error) {
	grep, err := api.AddParam(libSource("GrepPairMatched"), "opts", OptionsToString(*opts))
	if err != nil {
		return nil, err
	}

	return api.MapPartitionsWithIndex[string, string](input, grep)
}

func Grep(input *api.IDataFrame[string], o *SeqKitGrepOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitGrepOptions{}
	}
	opts := o.inner
	opts.setDefaults()
	aux := false
	opts.Count = &aux

	results, err := commonGrep(input, &opts)
	if err != nil {
		return nil, err
	}

	if *opts.BySeq && *opts.MaxMismatch > 0 {
		return results, nil
	} else if *opts.DeleteMatched && !*opts.InvertMatch {
		f1, err := api.Map[string, ipair.IPair[string, string]](results, libSource("GrepPairMatched"))
		if err != nil {
			return nil, err
		}
		f2, err := api.ToPair[string, string](f1).ReduceByKey(libSource("GrepReducePairMatched"), false)
		if err != nil {
			return nil, err
		}
		return api.Map[ipair.IPair[string, string], string](f2.FromPair(), libSource("GrepValueMatched"))
	} else {
		return results, nil
	}
}

func GrepCount(input *api.IDataFrame[string], o *SeqKitGrepOptions) (int64, error) {
	if o == nil {
		o = &SeqKitGrepOptions{}
	}
	opts := o.inner
	opts.setDefaults()
	aux := true
	opts.Count = &aux

	results, err := commonGrep(input, &opts)
	if err != nil {
		return 0, err
	}

	count, err := results.Reduce(libSource("GrepReduceCount"))
	if err != nil {
		return 0, err
	}

	return strconv.ParseInt(count, 10, 64)
}

File: bigseqkit-py/bigseqkit/grep.py
from typing import List

from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitGrepOptions:

    def __init__(self):
        self.__inner = GrepOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def pattern(self, v: List[str]):
        self.__inner.Pattern = v

    def patternFile(self, v: str):
        self.__inner.PatternFile = v

    def useRegexp(self, v: bool):
        self.__inner.UseRegexp = v

    def deleteMatched(self, v: bool):
        self.__inner.DeleteMatched = v

    def invertMatch(self, v: bool):
        self.__inner.InvertMatch = v

    def byName(self, v: bool):
        self.__inner.ByName = v

    def bySeq(self, v: bool):
        self.__inner.BySeq = v

    def onlyPositiveStrand(self, v: bool):
        self.__inner.OnlyPositiveStrand = v

    def maxMismatch(self, v: bool):
        self.__inner.MaxMismatch = v

    def ignoreCase(self, v: bool):
        self.__inner.IgnoreCase = v

    def degenerate(self, v: bool):
        self.__inner.Degenerate = v

    def region(self, v: str):
        self.__inner.Region = v

    def circular(self, v: bool):
        self.__inner.Circular = v

    def count(self, v: bool):
        self.__inner.Count = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()

        grep = _libSource("GrepPairMatched").addParam("opts", _optionsToString(opts))
        results = input.mapPartitionsWithIndex(grep)

        if opts.Count:
            count = results.reduce(_libSource("GrepReduceCount"))
            return int(count)
        elif opts.BySeq and opts.MaxMismatch > 0:
            return results
        elif opts.DeleteMatched and not opts.InvertMatch:
            f1 = results.mapPartitions(_libSource("GrepPairMatched"))
            f2 = f1.toPair().reduceByKey(_libSource("GrepReducePairMatched"), localReduce=False)
            return f2.map(_libSource("GrepValueMatched"))
        else:
            return results


class GrepOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.Pattern = None  # list[str]
        self.PatternFile = None  # str
        self.UseRegexp = None  # bool
        self.DeleteMatched = None  # bool
        self.InvertMatch = None  # bool
        self.ByName = None  # bool
        self.BySeq = None  # bool
        self.OnlyPositiveStrand = None  # bool
        self.MaxMismatch = None  # int
        self.IgnoreCase = None  # bool
        self.Degenerate = None  # bool
        self.Region = None  # str
        self.Circular = None  # bool
        self.Count = None  # bool

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "Pattern", [""])
        _setDefault(self, "PatternFile", "")
        _setDefault(self, "UseRegexp", False)
        _setDefault(self, "DeleteMatched", False)
        _setDefault(self, "InvertMatch", False)
        _setDefault(self, "ByName", False)
        _setDefault(self, "BySeq", False)
        _setDefault(self, "OnlyPositiveStrand", False)
        _setDefault(self, "MaxMismatch", 0)
        _setDefault(self, "IgnoreCase", False)
        _setDefault(self, "Degenerate", False)
        _setDefault(self, "Region", "")
        _setDefault(self, "Circular", False)
        _setDefault(self, "Count", False)


def grep(input: IDataFrame, o: SeqKitGrepOptions = None, **kwargs):
    if o is None:
        o = SeqKitGrepOptions()
    return o._run(input, **kwargs)
